src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlRetry/SqlRetryService.cs

```diff
@@ -383,5 +383,103 @@ public async Task TryLogEvent(string process, string status, string text, DateTi
+                    else if (replicaTrafficRatio > 0.99)
+                    {
+                        conn = await sqlConnectionBuilder.GetReadOnlySqlConnectionAsync(initialCatalog: null, cancellationToken: cancel).ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        var useWriteConnection = unchecked(Interlocked.Increment(ref _usageCounter)) % (int)(1 / (1 - _replicaTrafficRatio)) == 1; // examples for ratio -> % divider = { 0.9 -> 10, 0.8 -> 5, 0.75 - 4, 0.67 - 3, 0.5 -> 2, <0.5 -> 1}
+                        conn = useWriteConnection
+                                ? await sqlConnectionBuilder.GetSqlConnectionAsync(initialCatalog: null, cancellationToken: cancel).ConfigureAwait(false)
+                                : await sqlConnectionBuilder.GetReadOnlySqlConnectionAsync(initialCatalog: null, cancellationToken: cancel).ConfigureAwait(false);
+                    }
+                }
+
+                // Connection is never opened by the _sqlConnectionBuilder but RetryLogicProvider is set to the old, deprecated retry implementation. According to the .NET spec, RetryLogicProvider
+                // must be set before opening connection to take effect. Therefore we must reset it to null here before opening the connection.
+                conn.RetryLogicProvider = null; // To remove this line _sqlConnectionBuilder in healthcare-shared-components must be modified.
+                await conn.OpenAsync(cancel);
+
+                return conn;
+            }
+
+            private double GetReplicaTrafficRatio(ISqlConnectionBuilder sqlConnectionBuilder)
+            {
+                const int trafficRatioCacheDurationSec = 600;
+                if (_lastUpdated.HasValue && (DateTime.UtcNow - _lastUpdated.Value).TotalSeconds < trafficRatioCacheDurationSec)
+                {
+                    return _replicaTrafficRatio;
+                }
+
+                lock (_databaseAccessLocker)
+                {
+                    if (_lastUpdated.HasValue && (DateTime.UtcNow - _lastUpdated.Value).TotalSeconds < trafficRatioCacheDurationSec)
+                    {
+                        return _replicaTrafficRatio;
+                    }
+
+                    _replicaTrafficRatio = GetReplicaTrafficRatioFromDatabase(sqlConnectionBuilder);
+                    _lastUpdated = DateTime.UtcNow;
+                }
+
+                return _replicaTrafficRatio;
+            }
+
+            private static double GetReplicaTrafficRatioFromDatabase(ISqlConnectionBuilder sqlConnectionBuilder)
+            {
+                try
+                {
+                    using var conn = sqlConnectionBuilder.GetSqlConnection();
+                    conn.RetryLogicProvider = null;
+                    conn.Open();
+                    using var cmd = new SqlCommand("IF object_id('dbo.Parameters') IS NOT NULL SELECT Number FROM dbo.Parameters WHERE Id = 'ReplicaTrafficRatio'", conn);
```

Comment:
Do we have anywhere we are documenting these Parameter table values? There are now a few flags we can control from there and I think we should document them in a common location.